### PR TITLE
Lmdb storage backend

### PR DESCRIPTION
**Pull request by hakunamtu**
This pull request is addressing issue of somewhat high disk usage by NetDb and PeerProfiles, which is especially bad for devices with low flash and/or RAM (when using tmpfs), like cheap routers.
Instead of old filesystem-based storage [lmdb](https://symas.com/lmdb/technical/) simple key-value storage database is used. To preserve backwards-compatibility option added to switch between lmdb and old HashedStorage, old one is default value.
Other variants for a database were gdbm(not suitable due to a viral license), tinycdb (djb's constant database implementation, due to immutability forces to recreate and write it on disk on each update, which is not good for flash-based storages), levelDb and tdb(both too complex for such a simple task).
Lmdb is ligtweight (~90K on x86_64) and allows to reduce disk usage 4 times and more compared to old HashedStorage.
I'm new to development and code licenses, there should probably be now a copy of OpenLDAP llicense distributed with i2pd source code and binary as lmdb license says.
There is also a cdb storage backend which is ready and I can make a pull request to merge it too  if needed.

### CODE REVIEW

**Comment by majestrate** [comment]
I like this feature.

File: Makefile [hunk 1]
```diff
@@ -18,6 +18,7 @@ USE_AVX		:= yes
 USE_STATIC	:= no
 USE_MESHNET	:= no
 USE_UPNP	:= no
+USE_LMDB	:= yes
```
Comment (majestrate) [inline Makefile]: Ideally this should default to `no` for now.

File: libi2pd/RouterInfo.cpp [hunk 2]
```diff
@@ -553,6 +558,20 @@ namespace data
 		return bufbe64toh (buf + size) > m_Timestamp;
 	}
 
+	const uint8_t * RouterInfo::LoadBuffer(const uint8_t *buf, size_t len) //TODO: error handling
```
Comment (majestrate) [inline libi2pd/RouterInfo.cpp]: what does LoadBuffer do?

File: libi2pd/NetDb.cpp [hunk 3]
```diff
@@ -37,8 +37,14 @@ namespace data
 
 	void NetDb::Start ()
 	{
-		m_Storage.SetPlace(i2p::fs::GetDataDir());
-		m_Storage.Init(i2p::data::GetBase64SubstitutionTable(), 64);
+		std::string engine;
+		i2p::config::GetOption("storage.engine", engine);
+		if (engine == "lmdb")
```
Comment (majestrate) [inline libi2pd/NetDb.cpp]: this should have an `#ifdef LMDB`

File: libi2pd/RouterInfo.h [hunk 4]
```diff
@@ -120,6 +120,7 @@ namespace data
 			RouterInfo (const std::string& fullPath);
 			RouterInfo (const RouterInfo& ) = default;
 			RouterInfo& operator=(const RouterInfo& ) = default;
+			RouterInfo (const uint8_t * buf, int len, bool);
```
Comment (majestrate) [inline libi2pd/RouterInfo.h]: this should be

    RouterInfo(const uint8_t * buf, int len, bool checkSignature=true);

File: libi2pd/Storage.h [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	virtual ~FsIdentStorage();
+
+private:
+	i2p::fs::HashedStorage m_Storage;
+	std::string m_Fprefix;
+	bool m_IsB32;
+};
+
+#ifdef LMDB
+class MdbIdentStorage : public IdentStorage
```
Comment (majestrate) [inline libi2pd/Storage.h]: question: is MdbIdentStorage multithread safe?

**Comment by orignal** [comment]
Additional  dependencies are not allowed.

File: libi2pd/RouterInfo.cpp (hunk 2 above)
Comment (hakunamtu) [reply libi2pd/RouterInfo.cpp]: Basically, it copies buf to m_Buffer and calls RouterInfo::LoadBuffer without checking signature, like when we are loading from file.

File: libi2pd/Storage.h (hunk 5 above)
Comment (hakunamtu) [reply libi2pd/Storage.h]: No, when you call BeginUpdate or EndUpdate simultaneously from multiple threads, there will be a race condition.

**Comment by hakunamtu** [comment]
So, lmdb cannot be used as a storage. Could some changes be of any use? Should be a simple (maybe constant) key-value database implemented from scratch?

**Comment by orignal** [comment]
@hakunamtu, please explain what exactly you are trying to achieve and how. 
What do you need lmdb for it? Is it a way to reach the same result without it (for example using mmap)?

**Comment by l-n-s** [comment]
@hakunamtu addressing disk issues is a good idea, we have discussed it earlier in IRC, but it needs more discussion before we decide to pull in another dependency. 

Now there is a separate issue for it #1141

**Comment by orignal** [comment]
In general, additional dependancy could be added (for example libminiupnpc), but it must a clear reason for it.

**Comment by hakunamtu** [comment]
@original. the main idea is to reduce disk space used by i2pd databases, like NetDb and peerProfiles.  Now each record is stored in a separate file, and each file has a minimum size - in a case of tmpfs it's a PAGE_SIZE, 4K. Most of records are much smaller, for example - a peerProfiles with ~5.5M apparent size really uses 105M of disk space. This is especially hurtful for routers, where RAM and flash are limited, so it would be better to use more efficient storage. Like I said, I'm new to development, and thought that it would be better to bring an implementation.

**Comment by orignal** [comment]
@hakunamtu  so this PR solves the problem of disk space only?

**Comment by hakunamtu** [comment]
@original. yes. There could be less separate disk writes due to transactional nature of lmdb, but amount of information written is pretty much the same, I believe. More efficient serialization of router info and profile could help to reduce it.

**Comment by LLE8** [comment]
Is LMDB ported for OpenWrt?

**Comment by orignal** [comment]
@LLE8, I doubt we will accept this change anyway due unnecessary extra dependency

**Comment by LLE8** [comment]
I didn't found LMDB in OpenWRT, but change like this is interesting and very usefull for i2pd running in OpenWrt. Dependency on small database (liblmdb.so is about 80 KB in Debian) is not crytical, i think.
But LMDB missing in OpenWRT, it seems.

**Comment by hakunamtu** [comment]
@LLE8. I haven't found it in OpenWrt package base. There is another option - tinycdb, which is packaged, and have the smallest footprint from all of tested libs both in on-disk filesize and shared library size. However, it's constant, so it should be recreated every time you need to update database. That's terrible for flash, but it could be used with tmpfs. Another possible way is to increase an interval between flushing changes in NetDb on disk, maybe add a config option for it. Or maybe implement a database from scratch, using some properties of data specific for i2pd, like constant key size, to make job simpler.

**Comment by LLE8** [comment]
There are another alternative options:
1st
Trivial database (or New Trivial database) 
https://github.com/openwrt/packages/tree/master/libs/tdb
2nd
GDBM
https://github.com/openwrt/packages/tree/master/libs/gdbm

**Comment by hakunamtu** [comment]
Tdb seems like a good option. Gdbm looks not so good because, as I understand, you should license all the code under GPL even if you are just linking against it.

**Comment by LLE8** [comment]
In Debian 8
Trivial DB: libtdb.so is about 84 KB
New Trivial DB: libntdb.so is about 84 KB
GDBM: libgdbm.so is about 26 KB

**Comment by LLE8** [comment]
As i understand, if you DYNAMICally linking, you should NO license all the code under GPL.

**Comment by LLE8** [comment]
How many GPL libs?
```
~# ldd /usr/sbin/i2pd
        linux-vdso.so.1 (0x00007fffb45d5000)
        libcrypto.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007f7912c80000)
        libssl.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libssl.so.1.0.0 (0x00007f7912a16000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f79127fa000)
        libboost_system.so.1.55.0 => /usr/lib/x86_64-linux-gnu/libboost_system.so.1.55.0 (0x00007f79125f6000)
        libboost_date_time.so.1.55.0 => /usr/lib/x86_64-linux-gnu/libboost_date_time.so.1.55.0 (0x00007f79123e4000)
        libboost_filesystem.so.1.55.0 => /usr/lib/x86_64-linux-gnu/libboost_filesystem.so.1.55.0 (0x00007f79121cb000)
        libboost_program_options.so.1.55.0 => /usr/lib/x86_64-linux-gnu/libboost_program_options.so.1.55.0 (0x00007f7911f53000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f7911d36000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f7911a2a000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f7911729000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f7911513000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f7911167000)
        /lib64/ld-linux-x86-64.so.2 (0x000056035ddf8000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f7910f63000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f7910d5b000)
```

**Comment by LLE8** [comment]
https://edoras.sdsu.edu/doc/gdbm/gdbm.html

**Comment by radfish** [comment]
How stable/tested/maintained are these less known ones? Are they ACID/crash consistent? LMDB and Berkeley DB (BDB) are production quality. LMDB is also comparatively elegant in its CoW mechanism for guaranteeing crash consistency; and overall it's nicely simple in its design. There's a video of a talk about it. Nobody would want for this component to turn into a weak link in i2pd.

**Comment by LLE8** [comment]
LMDB has been ported for OpenWrt. Tested in LEDE 17.01.4 on TP-Link TL-WDR4300.

This is BETA! Full of BUGS!

Makefile:
```
# Copyright (C) 2006-2010 OpenWrt.org
#
# This is free software, licensed under the GNU General Public License v2.
# See /LICENSE for more information.
#

##LMDB for LEDE 17.01.4 LLE8

include $(TOPDIR)/rules.mk
PKG_NAME:=lmdb
PKG_VERSION:=0.9.21
PKG_RELEASE:=1

PKG_SOURCE:=LMDB_$(PKG_VERSION).tar.gz
PKG_SOURCE_URL:=https://github.com/LMDB/lmdb/archive/
PKG_HASH:=1187b635a4cc415bb6972bba346121f81edd996e99b8f0816151d4090f90b559
PKG_SOURCE_SUBDIR:=lmdb-LMDB_$(PKG_VERSION)
PKG_BUILD_DIR:=$(BUILD_DIR)/lmdb-LMDB_$(PKG_VERSION)
MAKE_PATH:=libraries/liblmdb

include $(INCLUDE_DIR)/package.mk

define Package/liblmdb0
        SECTION:=libs
        CATEGORY:=Libraries
        TITLE:=Lightning Memory-Mapped Database
        URL:=https://symas.com/lmdb/
        MAINTAINER:=X3_nobodY
endef

define Package/liblmdb0/description
        Lightning Memory-Mapped Database.
endef

define Package/lmdb-utils
        SECTION:=utils
        CATEGORY:=Utilities
        TITLE:=Lightning Memory-Mapped Database
        URL:=https://symas.com/lmdb/
        MAINTAINER:=X3_nobodY
endef

define Package/lmdb-utils/description
        Lightning Memory-Mapped Database.
endef

define Package/lmdb-test
        SECTION:=utils
        CATEGORY:=Utilities
        TITLE:=Lightning Memory-Mapped Database
        URL:=https://symas.com/lmdb/
        MAINTAINER:=X3_nobodY
endef

define Package/lmdb-test/description
        Lightning Memory-Mapped Database.
endef


define Build/Compile
        $(MAKE) -C $(PKG_BUILD_DIR)/$(MAKE_PATH) \
                CC="$(TARGET_CC)" \
                CFLAGS+="$(TARGET_CFLAGS)" \
                LDFLAGS+="$(TARGET_LDFLAGS)"
endef


define Build/InstallDev
        $(INSTALL_DIR) $(1)/usr/include
        $(INSTALL_DIR) $(1)/usr/lib
        $(CP) $(PKG_BUILD_DIR)/libraries/liblmdb/lmdb.h $(1)/usr/include
        $(CP) $(PKG_BUILD_DIR)/libraries/liblmdb/liblmdb.a $(1)/usr/lib
endef

define Package/liblmdb0/install
        $(INSTALL_DIR) $(1)/usr/lib
        $(CP) $(PKG_BUILD_DIR)/libraries/liblmdb/*.so*  $(1)/usr/lib
endef

define Package/lmdb-utils/install
        $(INSTALL_DIR) $(1)/usr/bin
        $(INSTALL_BIN) $(PKG_BUILD_DIR)/libraries/liblmdb/mdb_stat $(1)/usr/bin
        $(INSTALL_BIN) $(PKG_BUILD_DIR)/libraries/liblmdb/mdb_copy $(1)/usr/bin
        $(INSTALL_BIN) $(PKG_BUILD_DIR)/libraries/liblmdb/mdb_dump $(1)/usr/bin
        $(INSTALL_BIN) $(PKG_BUILD_DIR)/libraries/liblmdb/mdb_load $(1)/usr/bin
endef

define Package/lmdb-test/install
        $(INSTALL_DIR) $(1)/usr/bin
        $(INSTALL_BIN) $(PKG_BUILD_DIR)/libraries/liblmdb/mtest $(1)/usr/bin
endef

$(eval $(call BuildPackage,liblmdb0))
$(eval $(call BuildPackage,lmdb-utils))
$(eval $(call BuildPackage,lmdb-test))
```


Patch, copied from Debian, may be not required:
```
From: Debian Berkeley DB Group <pkg-db-devel@lists.alioth.debian.org>
Date: Tue, 22 Dec 2015 09:36:38 +0100
Subject: Debian quirks to make shared library versioned

---
 libraries/liblmdb/Makefile | 21 ++++++++++++---------
 1 file changed, 12 insertions(+), 9 deletions(-)

diff --git a/libraries/liblmdb/Makefile b/libraries/liblmdb/Makefile
index f254511..ddb6d91 100644
--- a/libraries/liblmdb/Makefile
+++ b/libraries/liblmdb/Makefile
@@ -27,10 +27,11 @@ CFLAGS      = $(THREADS) $(OPT) $(W) $(XCFLAGS)
 LDLIBS =
 SOLIBS =
 SOEXT  = .so
-prefix = /usr/local
+SOVER   = 0
+prefix = /usr
 exec_prefix = $(prefix)
 bindir = $(exec_prefix)/bin
-libdir = $(exec_prefix)/lib
+libdir = $(exec_prefix)/lib/$(DEB_HOST_MULTIARCH)
 includedir = $(prefix)/include
 datarootdir = $(prefix)/share
 mandir = $(datarootdir)/man
@@ -38,7 +39,7 @@ mandir = $(datarootdir)/man
 ########################################################################

 IHDRS  = lmdb.h
-ILIBS  = liblmdb.a liblmdb$(SOEXT)
+ILIBS  = liblmdb.a liblmdb$(SOEXT) liblmdb$(SOEXT).$(SOVER) liblmdb$(SOEXT).$(SOVER).0.0
 IPROGS = mdb_stat mdb_copy mdb_dump mdb_load
 IDOCS  = mdb_stat.1 mdb_copy.1 mdb_dump.1 mdb_load.1
 PROGS  = $(IPROGS) mtest mtest2 mtest3 mtest4 mtest5
@@ -49,13 +50,13 @@ install: $(ILIBS) $(IPROGS) $(IHDRS)
        mkdir -p $(DESTDIR)$(libdir)
        mkdir -p $(DESTDIR)$(includedir)
        mkdir -p $(DESTDIR)$(mandir)/man1
-       for f in $(IPROGS); do cp $$f $(DESTDIR)$(bindir); done
-       for f in $(ILIBS); do cp $$f $(DESTDIR)$(libdir); done
-       for f in $(IHDRS); do cp $$f $(DESTDIR)$(includedir); done
-       for f in $(IDOCS); do cp $$f $(DESTDIR)$(mandir)/man1; done
+       for f in $(IPROGS); do cp -pP $$f $(DESTDIR)$(bindir); done
+       for f in $(ILIBS); do cp -pP $$f $(DESTDIR)$(libdir); done
+       for f in $(IHDRS); do cp -pP $$f $(DESTDIR)$(includedir); done
+       for f in $(IDOCS); do cp -pP $$f $(DESTDIR)$(mandir)/man1; done

 clean:
-       rm -rf $(PROGS) *.[ao] *.[ls]o *~ testdb
+       rm -rf $(PROGS) *.[ao] *.[ls]o *.[ls]o.* *~ testdb

 test:  all
        rm -rf testdb && mkdir testdb
@@ -66,7 +67,9 @@ liblmdb.a:    mdb.o midl.o

 liblmdb$(SOEXT):       mdb.lo midl.lo
 #      $(CC) $(LDFLAGS) -pthread -shared -Wl,-Bsymbolic -o $@ mdb.o midl.o $(SOLIBS)
-       $(CC) $(LDFLAGS) -pthread -shared -o $@ mdb.lo midl.lo $(SOLIBS)
+       $(CC) $(LDFLAGS) -fPIC -Wl,-soname,liblmdb$(SOEXT).$(SOVER) -pthread -Wl,-Bsymbolic -shared -o $@.$(SOVER).0.0 mdb.lo midl.lo $(SOLIBS)
+       ln -s $@.$(SOVER).0.0 liblmdb$(SOEXT).$(SOVER)
+       ln -s liblmdb$(SOEXT).$(SOVER) liblmdb$(SOEXT)

 mdb_stat: mdb_stat.o liblmdb.a
 mdb_copy: mdb_copy.o liblmdb.a
```

**Comment by LLE8** [comment]
> How stable/tested/maintained are these less known ones?

TDB/NTDB are parts of Samba.
GDBM is standard UNIX 'dbm' analog.

**Comment by LLE8** [comment]
How to test it storage backend in OpenWrt?

**Pull request closed without merging**